src/mono/mono/mini/aot-compiler.c

```diff
@@ -14355,6 +14382,50 @@ add_interp_in_wrapper_for_sig (MonoAotCompile *acfg, MonoMethodSignature *sig)
 	add_method (acfg, wrapper);
 }
 
+static gboolean
+mono_aot_add_direct_pinvoke (MonoAotCompile *acfg, char *dpi)
+{
+	char *direct_pinvoke = g_strdup (dpi);
+	if (!direct_pinvoke)
+		return TRUE;
+
+	if (direct_pinvoke[0] == '#')
+		return TRUE;
+
+	g_strstrip (direct_pinvoke);
+	if (!direct_pinvoke)
+		return TRUE;
+
+	char **direct_pinvoke_split = g_strsplit (direct_pinvoke, "!", -1);
+	char *library_name = g_strdup (direct_pinvoke_split[0]);
+	if (!direct_pinvoke_split[1]) {
+		g_hash_table_insert (acfg->direct_pinvokes, library_name, NULL);
+		g_strfreev (direct_pinvoke_split);
+		return TRUE;
+	}
+	if (direct_pinvoke_split[2]) {
+		aot_printerrf (acfg, "The provided 'direct_pinvoke' AOT option '%s' shouldn't contain multiple '!'\n", dpi);
+		g_strfreev (direct_pinvoke_split);
+		return FALSE;
+	}
+
+	char *entrypoint_name = g_strdup (direct_pinvoke_split[1]);
+	if (g_hash_table_contains (acfg->direct_pinvokes, library_name)) {
+		GHashTable *val = g_hash_table_lookup (acfg->direct_pinvokes, library_name);
+		if (!val || g_hash_table_contains (val, entrypoint_name)) {
+			g_strfreev (direct_pinvoke_split);
+			return TRUE;
+		}
+		g_hash_table_insert (val, entrypoint_name, NULL);
+	} else {
+		GHashTable *val = g_hash_table_new_full (NULL, g_str_equal, g_free, NULL);
```

Comment:
Allocation failure.

Reply:
I assume you would like to use g_str_hash has hash function and not comparing memory pointers?

Reply:
Changed to use `g_str_hash`. I'm not familiar, why would this lead to an allocation failure?

Reply:
The allocation failure was related to calling g_hash_table_new_full, since it can alloc it can fail.

Reply:
Should this be handled by checking `if (!val)` and printing an error/cleaning up?
Is the problem that there is an assumption that val exists `g_hash_table_insert (val, entrypoint_name, NULL);` and we would need to avoid inserting into a non existent HashTable?
I'm wondering where there seems to be no other allocation failure handling for the other `g_hash_table_new_full`, or I just haven't found them?

In the same line of thinking, how often does `g_hash_table_new_full` and `g_strdup` fail, it doesn't seem like there are many checks to verify that they succeeded in other places in aot-compiler.c. Just wondering since the failure checks are making the logic a bit noisy (where we might even split the parsing into a separate helper function https://github.com/dotnet/runtime/pull/79721#discussion_r1051976500) compared to the Native AOT parallel https://github.com/dotnet/runtime/blob/ab2d195f169a08bf8eb978b472b2d3aca61b0a46/src/coreclr/tools/aot/ILCompiler/ConfigurablePInvokePolicy.cs#L44-L78

Reply:
C# have exceptions for all error handling, including OOM's, C don't so you need to handle all errors based on return values, could makes it more noisy but that mainly boils down to how you organize your code. The allocations fails when underlying malloc fails, normally due to OOM scenarios, on Mono its not strictly required in cases where allocation ends up in g_malloc or similar since it will do g_error that normally abort process in case of failure, but that is configuration based and could change, but as you said, the codebase doesn't always check returns for these allocations, due to assuming g_malloc and friends will abort process on OOM (default behavior). Normally its good practice writing solid code to handle memory allocation errors without aborting the process since OOM's can be temporary (especially on some platforms).

Reply:
That makes sense. I added error handling in case the HashTable could not be allocated, and this error bubbles up and causes the mono aot compiler to exit. We wouldn't want the compiler to proceed if a direct pinvoke that should be added fails to get added, right?